### DISCUSSION TITLE
Ambient Occlusion

### DISCUSSION THREAD

**Discussion by Vertexwahn** [mitsuba-renderer/mitsuba3#373] [Feature requests]
Is there an ambient occlusion integrator in Mitsuba3?

**Comment by Speierers** [comment]
Hi @Vertexwahn ,

Following this [tutorial](https://mitsuba.readthedocs.io/en/stable/src/rendering/scripting_renderer.html), it shouldn't be too hard to implement an ambient occlusion integrator plugin directly in Python.

**Reply by LCM1999** [reply]
Is there any tutorial for implement an ambient occlusion plugin in C++?

**Reply by njroussel** [reply]
Hi @LCM1999 

No, to my knowledge, there isn't.

**Comment by Yishun99** [comment]
Hi @Speierers ,
Is it possible to adapt the ambient occlusion to a differentiable version. 
I don't know how to get the gradient of visibility when computing ambient occlusion?

**Reply by njroussel** [reply]
Hi @DouYishun 

Yes, I believe it is, it would require a bit of work though.

In our existing integrators, we make use of reparameterization to handle the discontinuities introduces by some changing visibility. Have a look at the `emission_reparam`'s [implementation](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/emission_reparam.py). I think this is the easiest example of reparameterization we have. This is an advanced topic, I'd strongly suggest you read [some of the literature on it](https://people.csail.mit.edu/sbangaru/projects/was-2020/).
